core/src/main/java/org/apache/druid/math/expr/ApplyFunction.java

```diff
@@ -0,0 +1,611 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.math.expr;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import it.unimi.dsi.fastutil.objects.Object2IntArrayMap;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import org.apache.druid.java.util.common.IAE;
+import org.apache.druid.java.util.common.RE;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public interface ApplyFunction
+{
+  String name();
+
+  ExprEval apply(LambdaExpr lambdaExpr, List<Expr> argsExpr, Expr.ObjectBinding bindings);
```

Comment:
I "think" there are variable name scoping issues in all implementations . one option is of course to fix all of it. however, given that this is meant to be a simple thing and we would like to keep the impl simple, we can just enforce the rule all arg variable names (whether or not  in lambda expressions) must be unique and verify that at time of expression parsing . I think even with that rule, language would be allow all intended use cases.

Reply:
I'm willing to fix any scoping issues if you notice them, the place to look is `Expr.ObjectBinding` implementations such as `SettableLambdaBinding`, `IndexableMapLambdaObjectBinding`, etc. The basic process for all of them is that the `ApplyFunction` argument expressions are evaluated against the input `Expr.ObjectBinding` and then the input `Expr.ObjectBinding` is wrapped with a `Expr.ObjectBinding` implementation that intercepts lambda identifiers to provide the evaluated arguments instead. I've done quite a lot of testing re-using the same variables in a nested manner and things seem to have worked out ok so far, and a couple of tests in `ApplyFunctionTest.java` and maybe `ParserTest.java` cover at least some of these cases, but I will add some more tests and try to break it.

Reply:
took a closer look now, it seems ok.